connect/clients/nats.py

```diff
@@ -153,6 +184,102 @@ def nats_timing_event_handler(msg: Msg):
+
+    try:
+        # attempt to retransmit the message
+        logger.trace(
+            f"do_retransmit #{message['retransmit_count']}: retransmitting to: {message['target_endpoint_url']}"
+        )
+        async with AsyncClient(verify=settings.certificate_verify) as client:
+            if message["operation"] == "POST":
+                await client.post(message["target_endpoint_url"], json=resource)
+            elif message["operation"] == "PUT":
+                await client.put(message["target_endpoint_url"], json=resource)
+            elif message["operation"] == "PATCH":
+                await client.patch(message["target_endpoint_url"], json=resource)
+
+        # if the message came from the retransmit queue, remove it
+        if not queue_pos == -1:
+            nats_retransmit_queue.pop(queue_pos)
+        message["status"] = "SUCCESS"
+        logger.trace(
+            f"do_retransmit: successfully retransmitted message with id {message['uuid']} "
+            + f"after {message['retransmit_count']} retries"
+        )
+    except Exception as ex:
+        logger.trace(f"do_retransmit: exception {ex}")
+        if queue_pos == -1:
+            nats_retransmit_queue.append(message)
+            logger.trace(f"do_retransmit: queued message for retransmitter()")
+
+        if not max_retries == -1 and message["retransmit_count"] >= max_retries:
+            nats_retransmit_queue.pop(queue_pos)
+            message["status"] = "FAILED"
+            logger.trace(
+                f"do_retransmit: failed retransmit of message with id {message['uuid']} "
+                + f"after {message['retransmit_count']} retries"
+            )
+
+    # send outcome to kafka
+    if message["status"] == "SUCCESS" or message["status"] == "FAILED":
+        transmit_delta = datetime.now() - datetime.strptime(
+            message["transmit_start"], "%Y-%m-%dT%H:%M:%S.%f"
+        )
+        message["elapsed_transmit_time"] = transmit_delta.total_seconds()
+        message["elapsed_total_time"] += transmit_delta.total_seconds()
+        await kafka_producer.produce(
+            "RETRANSMIT", json.dumps(message, cls=ConnectEncoder)
+        )
+        logger.trace(f"do_retransmit: sent message to kafka topic RETRANSMIT")
+
+
+async def retransmitter():
+    """
+    Process messages in nats_retransmit_queue.
+    """
+    global nats_retransmit_queue
+    settings = get_settings()
+    logger.trace("Starting retransmit loop")
+
+    while not nats_retransmit_canceled:
+        # iterate through nats_retransmit_queue
+        for i in range(len(nats_retransmit_queue)):
+            logger.trace(
+                f"retransmitter: retransmitting from nats_retransmit_queue position {i}"
+            )
+            await do_retransmit(nats_retransmit_queue[i], i)
+            if i + 1 < len(nats_retransmit_queue):
+                await asyncio.sleep(2)
+        await asyncio.sleep(settings.nats_retransmit_loop_interval_secs)
+
```

Comment:
retransmitter() is running as an infinite loop on the main event loop, so taking care to make it well-behaved *and* not monopolize the do_transmit() function which is also used to process messages coming from NATS.